daffodil-io/src/main/scala/org/apache/daffodil/io/DirectOrBufferedDataOutputStream.scala

```diff
@@ -802,89 +924,29 @@ class DirectOrBufferedDataOutputStream private[io] (var splitFrom: DirectOrBuffe
-  // need for a bufferingJOS that the parent class would allocate. However, the
-  // bufferJOS is used to determine if a DOS is buffering or not. In our case,
-  // this is buffering if the JavaOutputStream is the same as bufferingJOS
-  // (i.e. null).
-  override protected val bufferingJOS = null
-
-  // This function is exactly the same as the one it overrides in the parent
-  // except that it does not have the assertion that _javaOutputStream is never
-  // null. This is because when this BLOB DOS is not direct, the
-  // javaOutputStream is set to null (and should never be written to). Once
-  // this becomes direct, the _javaOutputStream variable is changed to the
-  // direct OutputStream.
-  override def getJavaOutputStream() = {
-     _javaOutputStream
-   }
-
-  protected override def deliverContent(directDOS: DirectOrBufferedDataOutputStream, finfo: FormatInfo) = {
-
-    val blobStream =
-      try {
-        Files.newInputStream(path, StandardOpenOption.READ)
-      } catch {
-        case e: Exception =>
-          throw new BlobIOException(
-            "Unable to open BLOB %s for reading: %s".format(
-              path.toString,
-              e.getMessage()))
-      }
-
-    val array = new Array[Byte](blobChunkSizeInBytes)
-
-    val blobChunkSizeInBits = blobChunkSizeInBytes * 8
-
-    var remainingBitsToPut = lengthInBits
-    var fileHasData = true
-    while (remainingBitsToPut > 0 && fileHasData) {
-      val bitsToRead = Math.min(remainingBitsToPut, blobChunkSizeInBits)
-      val bytesToRead = (bitsToRead + 7) / 8
-      val bytesRead = blobStream.read(array, 0, bytesToRead.toInt)
-      if (bytesRead == -1) {
-        fileHasData = false
-      } else {
-        val bitsToPut = Math.min(bytesRead * 8, bitsToRead)
-        val ret = directDOS.putByteArray(array, bitsToPut.toInt, finfo)
-        if (!ret) {
-          blobStream.close()
-          throw new BlobIOException(
-            "Failed to write BLOB data: %s".format(path.toString))
+          Assert.invariant(directDOS.putLongUnchecked(
+            fragNum, nFragBits, finfo,
+            ignoreByteOrder=this.bufferingJOS.isFile))
         }
-        remainingBitsToPut -= bitsToPut
       }
     }
+    bufOS.close()
+  }
 
-    blobStream.close()
-
-    // calculate the skip bits
-    val nFillBits = remainingBitsToPut
-    if (nFillBits > 0) {
-      val ret = directDOS.skip(nFillBits, finfo)
-      if (!ret) {
-          throw new BlobIOException(
-            "Failed to skip %s bits after BLOB data: %s".format(nFillBits, path.toString))
-      }
+  /**
+   * Clean up any temporary files that were generated
+   */
+  def cleanUp(): Unit = {
+    var maybeStream = Maybe(this)
+    while (maybeStream.isDefined) {
+      val stream = maybeStream.get
+      val next = stream._following
+      if (stream.bufferingJOS.isTempFile && stream.bufferingJOS.maybeFile.isDefined)
+        stream.bufferingJOS.maybeFile.get.delete()
+      maybeStream = next
```

Comment:
I'm not sure if this really matters, but I think this will do a double delete.

1. Say we have a direct DOS and we split off to a buffered DOS.
2. Assume the buffered DOS switches from byte-based to file-base.
3. No errors occur during the unparse and we deliver the buffered DOS to the direct DOS.
4. After the content is delievered, we delete the temp file and make the buffered DOS the new direct DOS.
    * Even though this DOS is direct, we still have a pointer to bufferingJOS which is still file-based (though the file has been deleted). This is potentially an existing bug--we probably do not want to hang on to the bufferingJOS once it's been delivered, but that is an already existing bug so probably not worth worrying about for this.
5. We call cleanup(), this DOS bufferingJOS is a temp file and maybeFile.defined, so we will delete the file. But we already deleted the file in step 4, so have a double delete.

Fortunately, I think FIle.delete just returns false when the file doesn't exist, so it might not really be a big deal, but it does sort of feel wrong to try to delete a file that we already deleted. I *think* the fix would be to only do a delete if stream.isBuffering, since a buffering stream that exists here was never delivering and so it was also neve deleted.

I might also recommend refactoring out the while loop and just walk the _following DOS's calling cleanUp, e.g.

```scala
def cleanUp(): Unit = {
  // logic to test if temp and delete if so

  if (_following.isDefined)
    _following.get.cleanUp()
}
```

Avoids the Maybe's and feels more scala-y.

Reply:
Perhaps I should have put in a comment, but I did expect it to do a double delete and was OK with that as File.delete will simply return false if the file does not exist.  It does not throw an error.

Will make the change to check for isBuffering and refactor out the while loop.